<a href="https://colab.research.google.com/github/jyj1206/Python/blob/main/Python_everyday_problems.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 중간고사

Day 0
> 숫자를 입력 받아서 짝수인지 홀수인지 출력하는 프로그램을 작성하세요

In [ ]:
n = int(input("정수를 입력하세요 : "))
if n % 2 == 0:
    print(n, "은(는) 짝수입니다")
else:
    print(n, "은(는) 홀수입니다")

정수를 입력하세요 : 4
4 은(는) 짝수입니다


Day 1

> 사용자로부터 두 개의 정수를 입력으로 받습니다.
두 개 정수의 합, 정수의 차, 정수의 곱, 정수의 평균, 큰 수, 그리고 작은 수를 계산하여 화면에 출력하는 파이썬 프로그램을 작성하세요.

In [ ]:
num1, num2 = map(int, input("두 정수를 입력하세요").split())

print("두 정수의 합", num1 + num2)
print("두 정수의 차", num1 - num2 if num1 > num2 else num2 - num1)
print("두 정수의 곱", num1 * num2)
print("두 정수의 평균", (num1 + num2)/2)
print("두 정수 중 큰 수", num1 if num1 > num2 else num2)
print("두 정수 중 작은수", num2 if num1 < num2 else num1)

Day 2
> 사용자로부터 4자리의 정수를 입력으로 받아서 자리수의 합을 계산하는 프로그램을 작성하세요.
예를 들면 사용자가 1357을 입력한다면 1+3+5+7를 계산하고 그 결과를 출력하세요.

In [ ]:
n = int(input("4자리 수를 입력하세요."))

total = n % 10
n //= 10
total += n % 10
n //= 10
total += n % 10
n //= 10
total += n % 10

print(total)

Day 3
> BMI(Body Mass Index)는 체중(Kg)을 신장(m)의 제곱으로 나눈 값이다.
BMI는 체지방 축적을 잘 반영하기 때문에 비만도 판정에 많이 사용한다.
BMI 값이 18.50보다 작을 때는 저체중,
BMI 값이 18.50에서 24.99일 때는 표준,
BMI 값이 25 이상이고 29.99이하 일 때는 과체중,
BMI 값이 30 이상일때는 비만으로 판정한다.
사용자로부터 신장과 체중을 입력으로 받아서 BMI 값을 출력하는 프로그램을 작성하세요.

In [ ]:
height = float(input("신장을 입력하세요(cm 단위) : "))
weight = float(input("체중을 입력하세요(kg 단위) : "))
BMI = weight / ((height/100) ** 2)
if BMI >= 30:
  print("비만")
elif BMI >= 25:
  print("과체중")
elif BMI >= 18.5:
  print("표준")
else:
  print("저체중")

Day 4
> 사용자로부터 2개의 정수를 입력으로 받습니다.
두 수가 같으면 1을, 다르면 -1을 출력하는 프로그램을 작성하세요.

In [ ]:
a, b = map(int, input("두 정수를 입력하세요 : ").split())

if a==b:
  print("1")
else:
  print("-1")

Day 5
> 각(angle)에서 0도 초과 90도 미만은 예각, 90도는 직각, 90도 초과 180도 미만은 둔각 180도는 평각으로 분류합니다.
사용자가 각을 입력했을 때, 예각이면 1, 직각이면 2, 둔각이면 3, 평각이면 4를 출력하는 프로그램을 작성하세요.
사용자의 입력은 0부터 180이라고 가정하자.

In [ ]:
angle = int(input("각을 입력하세요 : "))
if angle == 180:
  print("평각")
elif angle > 90:
  print("둔각")
elif angle == 90:
  print("직각")
else:
  print("예각")

Day 6
> 사용자로부터 2개의 정수를 입력으로 받습니다.
각각의 정수를 num1과 num2라고 한다.
num1이 num2의 배수일 경우에는 "num1은 num2의 배수입니다."라고 출력한다.
num1이 num2의 배수가 아닐 경우에는 "num1은 num2의 배수가 아닙니다."라고 출력한다.

In [ ]:
num1, num2 = map(int, input("2개의 정수를 입력하세요(단, 2번째 입력하는 수는 0이 아님) : ").split())
if (num1 % num2 == 0) :
    print(f"{num1}은(는) {num2}의 배수입니다.")
else :
    print(f"{num1}은(는) {num2}의 배수가 아닙니다.")

Day 7
> 새로 생 피자 가게는 피자를 일곱 조각으로 잘라 줍니다.
사용자로부터 피자를 나눠 먹을 사람의 수 n을 입력으로 받는다.
모든 사람이 피자를 한 조각 이상 먹기 위해 필요한 피자의 수를 계산하여 출력하는 프로그램을 작성하세요.

In [ ]:
from math import ceil
n = int(input("피자를 나눠 먹을 사람의 수 : "))
print(ceil(n/7))

Day 8
> 춘식이는 추운 겨울에도 아이스 아메리카노만 마시는 소위 말하는 얼죽아입니다.
아이스 아메리카노는 한 잔에 4,500원입니다.
춘식이가 가지고 있는 돈 money를 사용자로부터 입력으로 받습니다.
춘식이가 최대로 마실 수 있는 아이스 아메리카노의 잔 수와 남는 돈을 순서대로 담은 리스트를 출력하는 프로그램을 작성하세요.

In [ ]:
money = int(input("춘식이가 가지고 있는 돈을 입력하세요 : "))
print([money // 4500, money % 4500])

Day 9
> 정수를 담은 임의의 리스트 num_list가 있습니다.
가장 첫 번째 원소를 1번 원소라고 할 때,
홀수 번째 원소들의 합과 짝수 번째 원소들의 합을 계산합니다.
두 합 중에서 큰 값을 출력하는 프로그램을 작성하세요.

In [ ]:
num_list = list(range(1, 11)) # 임의의 리스트

# 홀수번째 원소 합
odd_sum = sum([num_list[i] for i in range(1, len(num_list), 2)])

# 짝수번째 원소 합
even_sum = sum([num_list[i] for i in range(0, len(num_list), 2)])

# 두 합 중 큰 값 출력
print(max(odd_sum, even_sum))

Day 10
> 10이하의 자연수에서 3 또는 5의 배수는, 3, 5, 6, 9이고 이 숫자들의 합은 23이다.
1000이하의 자연수에서 3 또는 5의 배수인 숫자들의 합을 구하는 프로그램을 작성하세요.

In [ ]:
total = 0
for n in range(1, 1001):
  if n % 3 == 0 or n % 5 == 0:
    total += n
print(total)

Day 11
> 피보나치 수열(Fibonacci sequence)은 각 항이 앞의 두 항의 합으로 이루어진 수열입니다.
사용자로부터 양의 정수 n을 입력으로 받아서 피보나치 수열의 n번째 항의 값을 출력하는 프로그램을 작성하세요.
단, 반복문을 이용하여 프로그램을 작성하세요.
피보나치 수열에 대한 자세한 사항은 아래의 내용을 참고하세요.

>피보나치 수열의 정의:
   - 첫 번째 항: 0
   - 두 번째 항: 1
   - 세 번째 항부터는 앞의 두 항의 합으로 계산됩니다.

> 즉, 피보나치 수열은 다음과 같습니다.
0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, ...
>
> - 첫 번째 항 F(0)=0
> - 두 번째 항 F(1)=1
> - 세 번째 항 F(2)=F(1)+F(0)=1+0=1
> - 네 번째 항 F(3)=F(2)+F(1)=1+1=2
> - 다섯 번째 항 F(4)=F(3)+F(2)=2+1=3

>피보나치 수열을 수식으로 나타내면 다음과 같습니다.
>- F(n)=F(n−1)+F(n−2) (단, n≥2)

In [ ]:
n = int(input("피보나치 수의 몇 번째 항인지 입력하세요.(단 n은 1부터 시작) : "))
fibo = [0] * (n + 1)
fibo[2] = 1
if n == 1 or  n == 2:
  print(fibo[n])
else:
  for i in range(3, n + 1):
    fibo[i] = fibo[i - 1] + fibo[i - 2]
  print(fibo[n])

Day 12
> 공백(white space)을 다수 포함한 문자열이 주어졌다.
공백을 모두 제거한 새로운 문자열을 만들어 보고자 한다.
원래 문자열과 새로운 문자열을 출력하는 프로그램을 작성하세요.

In [ ]:
s = input("공백을 다수 포함한 문자열 입력 : ")
print("원래 문자열 :", s)
print("새로운 문자열 :", s.replace(' ', ''))

공백을 다수 포함한 문자열 입력 : 아 아 아아
원래 문자열 : 아 아 아아
새로운 문자열 : 아아아아


Day 13
> <이상한 변호사 우영우?를 보셨나요?
주인공 우영우는 자기 소개를 할 때,
"똑바로 읽어도 거꾸로 읽어도 우영우입니다.
기러기, 인도인, 토마토, 스위스, 별똥별, 우영우"라고 말합니다.
똑바로 읽어도 거꾸로 읽어도 같은 문자열을 회문(回文)이라고 합니다.
영어로는 palindrome 이라고 합니다.
사용자로부터 입력을 받은 문자열이 회문(回文) / palindrome 인지 아닌지 판별하는 프로그램을 작성하세요.

In [ ]:
sentence = input("문자열을 입력하세요 : ")
if sentence == sentence[::-1]:
  print("회문입니다.")
else:
  print("회문이 아닙니다.")

문자열을 입력하세요 : 토마토
회문입니다.


Day 14
> 사용자로부터 두 개의 문자열을 입력으로 받습니다.
첫 번째 문자열 속에 두 번째 문자열이 존재하는지 여부를 판별하는 프로그램을 작성하세요.

In [ ]:
first_sentence = input("첫 번째 문자열을 입력하세요 : ")
second_sentence = input("두 번째 문자열을 입력하세요 : ")
if second_sentence in first_sentence:
  print("첫 번째 문자열 속에 두 번째 문자열이 존재합니다.")
else:
  print("첫 번째 문자열 속에 두 번째 문자열이 존재하지 않습니다.")

첫 번째 문자열을 입력하세요 : 사과 많이 많이
두 번째 문자열을 입력하세요 :  많이
첫 번째 문자열 속에 두 번째 문자열이 존재합니다.


Day 15
> "에라토스테네스(Eratosthenes)의 체"가 무엇일까요?
소수(Prime Number)를 구하는 방법 중에 하나입니다.
자세한 내용은 아래의 링크를 참조하시기 바랍니다.
"에라토스테네스의 체"를 활용하여, 1부터 10,000까지의 소수를 구해볼까요?
총 소수의 갯수와, 소수들을 나열하는 프로그램을 작성하세요.

In [ ]:
is_prime_list = [False, False] + [True] * 9999

for i in range(2, int(len(is_prime_list) ** 0.5) + 1):
  if is_prime_list[i]:
    for j in range(i + i, len(is_prime_list),i):
      is_prime_list[j] = False

print(sum(is_prime_list))
print([i for i, is_prime in enumerate(is_prime_list) if is_prime == True])

1229
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 121

Day 16
> 함수의 파라미터로 두 개의 문자열을 받습니다.
두 개의 문자열이 anagram이면 True를 반환하고, 아니면 False를 반환합니다.

> 사용자로부터 두 개의 문자열을 받아서, anagram 여부를 판별하는 프로그램을 작성하세요.

> Anagram의 예를 들면 다음과 같습니다.
1. listen - silent
2. race - care
3. eat - tea

In [ ]:
def is_anagram(str1, str2):
  # 대 소문자 구분 없음
  str1 = str1.lower()
  str2 = str2.lower()

  # 공백 제거
  str1 = str1.replace(" ", "")
  str2 = str2.replace(" ", "")

  # 두 문자열의 길이가 같지 않으면
  if len(str1) != len(str2):
    return False
  else :
    alpa = [0] * 26

    # 첫번째 문자열에 각 알파벳 갯수 저장
    for ch in str1:
      alpa[ord(ch) - ord('a')] += 1

    # 두번째 문자열과 비교
    for ch in str2:
      index = ord(ch) - ord('a')
      if alpa[index] <= 0:
        return False
      alpa[index] -= 1

    return True

string1 = input("첫 번째 문자열 : ")
string2 = input("첫 번째 문자열 : ")

if is_anagram(string1, string2):
  print("두 문자열은 Anagram입니다.")
else:
  print("두 문자열은 Anagram이 아닙니다.")

첫 번째 문자열 : Moon starer
첫 번째 문자열 : Astronomer
두 문자열은 Anagram입니다.


Day 17
> 함수를 만들어볼까요?
매개변수로 문자열을 받습니다.
문자열에 포함된 모음을 제거한 새로운 문자열을 반환합니다.
문자열은 영어 기준입니다.

> 사용자로부터 입력으로 받은 문자열을 인자로 사용하여 위의 함수를 호출합니다.
직접 만든 함수가 제대로 동작하는지 테스트 결과를 살펴보세요.

In [ ]:
def remove_vowels(s):
  vowels = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']

  new_s = ''
  for ch in s:
    if ch in vowels:
      continue
    new_s += ch
  return new_s

str_ = input("문자열을 입력하세요 : ")
print(f"모음을 제거한 새로운 문자열은 {remove_vowels(str_)} 입니다.")

문자열을 입력하세요 : string
모음을 제거한 새로운 문자열은 strng 입니다.


Day 18
> 함수를 만들어볼까요?
매개변수로 받은 문자열에서 가장 긴 단어를 찾아내서 반환해주는 함수입니다.

> 사용자로부터 한 문장(영어)을 문자열로 입력으로 받습니다.
그 문자열을 인자로 사용하여 만든 함수를 호출합니다.
본인이 만든 함수가 제대로 동작하는지 테스트해볼까요?

In [ ]:
def longest_word(s):
  words = s.split()
  longest = ''
  for word in words:
    if len(word) > len(longest):
      longest = word
  return longest

sentence = input("영어 문장을 입력하세요 : ")
print(f"문자열에서 가장 긴 단어는 {longest_word(sentence)}입니다.")

영어 문장을 입력하세요 : I love you
문자열에서 가장 긴 단어는 love입니다.


Day 19
> 회문문자(palindrome)을 판단하는 파이썬의 간결한 표현은 다음과 같습니다.
```python
is_palindrome = lambda s: s == s[::-1]
```
>  회문문자가 있다면 회문숫자가 있겠죠?
> 회문숫자는 앞뒤가 대칭인 숫자, 예를 들면, 9009입니다.
> 9009는 두 자리 숫자 두 개의 곱(91*99)을 통해서 만들 수 있는 가장 큰 회문숫자입니다.
> 세 자리의 숫자 두 개의 곱을 통해서 만들 수 있는 가장 큰 회문숫자를 찾는 프로그램을 만들어볼까요?
> 그 회문숫자를 구성하는 세 자리의 숫자 두 개는 무엇일까요?

In [ ]:
def is_palindrome(s):
  return s == s[::-1]

max_palindrome = 0
for i in range(999, 99, -1):
  for j in range(i, 99, -1):
    if is_palindrome(str(i * j)):
      if i * j > max_palindrome:
        max_palindrome = i * j
        max_pair = (i, j)

print(f"세자리 숫자의 두 개의 곱을 통해서 만들 수 있는 가장 큰 회문숫자는 {max_palindrome}입니다.")
print(f"그 회문 숫자를 구성하는 두 개는 {max_pair[0]}(와)과 {max_pair[1]}입니다.")

세자리 숫자의 두 개의 곱을 통해서 만들 수 있는 가장 큰 회문숫자는 906609입니다.
그 회문 숫자를 구성하는 두 개는 993(와)과 913입니다.


Day 20
> 반복문과 break 명령어를 활용하여 아래와 같은 결과를 출력하는 프로그램을 작성하세요.

```
#
# 1
# 1 2
# 1 2 3
# 1 2 3 4
# 1 2 3 4 5
```

In [ ]:
for i in range(6):
  print("#", end = ' ')
  for j in range(1, 6):
    if j > i:
      break
    print(j, end= ' ')
  print()

# 
# 1 
# 1 2 
# 1 2 3 
# 1 2 3 4 
# 1 2 3 4 5 


Day 21
> 정수 10개를 담고 있는 리스트가 있습니다.
값이 작은 정수부터 크기 순서로 차례대로 정렬한 리스트를 만드는 프로그램을 작성하세요.
Python 내부함수를 이용하여 리스트를 정렬하는 방법은 권하지 않습니다.

In [ ]:
def selection_sort(numbers):
  for i in range(9):
    min_index = i
    for j in range(i+1, 10):
      if numbers[min_index] > numbers[j]:
        min_index = j
    numbers[i], numbers[min_index] = numbers[min_index], numbers[i]

numbers = [9, 5, 4, 2, 3, 7, 8, 1, 10, 6]
selection_sort(numbers)
print(numbers)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


Day 22
> 영어 문장이 주어졌을 때, 특정 단어가 그 문장의 몇 번째 단어인지 확인하는 프로그램을 작성하세요.
예를 들면, "If music be the food of love, play on." 문장에서 "play"는 8번째 단어입니다.

In [ ]:
def find_word(sentence, word):
  replaced_sentence = sentence.replace("."," ").replace(",", " ").replace("!", " ").replace("?", " ") # 구두점 제거

  location = [w for w in replaced_sentence.lower().split()].index(word.lower()) + 1
  print(f'"{sentence}"문장에서 "{word}"는 {location}번쨰 단어입니다.')

sentence = input("영어 문장을 입력하세요 : ")
word = input("찾고자하는 단어를 입력하세요 : ")
find_word(sentence, word)

영어 문장을 입력하세요 : If music be the food of love, play on.
찾고자하는 단어를 입력하세요 : play
"If music be the food of love, play on."문장에서 "play"는 8번쨰 단어입니다.


Day 23
> "To be, or not to be, that is the question" 라는 문장을 들어보셨나요?
이 문장에는 단어 "be"가 두 차례 나타납니다.
사용자 입력으로 받은 문장에서 특정 단어가 몇 차례 나타나는 지를 알려주는 프로그램을 작성해보세요.

In [ ]:
def return_count_word(sentence, word):
  replaced_sentence = sentence.replace("."," ").replace(",", " ").replace("!", " ").replace("?", " ") # 구두점 제거

  words = replaced_sentence.lower().split()
  count_word = words.count(word.lower())
  return count_word

sentence = input("영어 문장을 입력하세요 : ")
word = input("찾고자하는 단어를 입력하세요 : ")
print(f'"{sentence}"문장에서 "{word}"는 {return_count_word(sentence, word)}번 나타납니다.')

영어 문장을 입력하세요 : To be, or not to be, that is the question
찾고자하는 단어를 입력하세요 : be
"To be, or not to be, that is the question"문장에서 "be"는 2번 나타납니다.


Day 24
> 영문자 a를 1, b를 2, c를 3이라고 하고, 이후에 모든 영문자에 숫자를 차례대로 부여하자.
사용자로부터 영어 단어를 하나 입력 받습니다.
그 단어에 포함된 모든 영어 단어에 해당하는 숫자의 값을 모두 더한다.
영어 단어와 함께 그 숫자들의 합을 출력하는 프로그램을 작성하세요.
이 프로그램을 작성할 때, 내장함수(built-in) 함수 ord( )를 참고하세요.

In [ ]:
word = input("영어 단어 하나를 입력하세요 (소문자) : ")
print(sum(map(lambda x : ord(x) - ord('a') + 1 , word)))

영어 단어 하나를 입력하세요 (소문자) : apple
50


Day 25
> 모든 자연수 x에 대해서 현재 값이 x이면 x가 짝수일 때는 2로 나누고, x가 홀수일 때는 3 * x + 1로 바꾸는 계산을 계속해서 반복하면 언젠가는 반드시 x가 1이 되는지 묻는 문제를 콜라츠 문제라고 부릅니다.
그리고 위 과정에서 거쳐간 모든 수를 기록한 수열을 콜라츠 수열이라고 부릅니다.
계산 결과 1,000 보다 작거나 같은 수에 대해서는 전부 언젠가 1에 도달한다는 것이 알려져 있습니다.

>임의의 1,000 보다 작거나 같은 양의 정수 n이 주어질 때 초기값이 n인 콜라츠 수열을 반환하는 함수 collatz를 완성해 주세요.
collatz 함수가 제대로 작동하는지 테스트하는 코드를 작성하세요.

>입력 n : 10
출력 result : [10, 5, 16, 8, 4, 2, 1]

In [ ]:
def return_collatz(n):
  result = [n]
  while n != 1:
    if n % 2 == 0:
      n //= 2
    else:
      n = 3 * n + 1
    result.append(n)

  return result

n = int(input("입력 n : "))
print("출력 result :", return_collatz(n))

입력 n : 3
출력 result : [3, 10, 5, 16, 8, 4, 2, 1]


Day 26
> 약수의 개수가 세 개 이상인 수를 합성수라고 합니다.

> 자연수 n이 매개변수로 주어질 때 n이하의 합성수의 개수를 반환하도록 함수를 구현하세요.

> 그 함수가 제대로 동작하는지 테스트하는 코드를 작성하세요.
사용자로부터 n을 입력으로 받고, 합성수의 개수를 출력합니다.
예를 들면, 15 이하 합성수는 4, 6, 8, 9, 10, 12, 14, 15 로 8개입니다.
따라서 8을 출력합니다

In [ ]:
def count_composite_number(n):
  count = 0
  # 소수가 아니면 합성수이므로, 단 1은 약수의 개수가 1개이므로 제외
  for i in range(2, n + 1):
    if is_prime(i):
      continue
    count += 1
  return count


def is_prime(n):
  if n == 1:
    return False
  for i in range(2, int(n ** 0.5) + 1):
    if n % i == 0:
      return False
  return True

n = int(input("자연수 n을 입력하세요 : "))
print(f"{n} 이하 합성수의 개수는 {count_composite_number(n)}개 입니다.")

자연수 n을 입력하세요 : 15
15 이하 합성수의 개수는 8개 입니다.


Day 27
> 재귀함수(recursive function)를 이용해서 문제를 해결할 수 있는 방법을 설명해드렸습니다.
정수 n과 m 사이의 모든 정수의 합을 구하는 문제도 재귀를 사용하여 풀 수도 있습니다.
조건은 n은 m보다 작은 정수라는 것입니다.

In [ ]:
def sum_numbers(n, m):
  if n == m:
    return n
  else:
    return n + sum_numbers(n + 1, m)

n = int(input("정수 n을 입력하세요 : "))
m = int(input("정수 m을 입력하세요 : "))
print(f"{n}과 {m}사이의 모든 정수의 합은 {sum_numbers(n, m)}입니다.")

정수 n을 입력하세요 : 4
정수 m을 입력하세요 : 5
4과 5사이의 모든 정수의 합은 9입니다.


Day 28 시험 주 생략

# 기말고사

Day 29
> 양의 정수 n이 매개변수로 주어질 때, n이 홀수라면 n 이하의 홀수인 모든 양의 정수의 합을 반환하고 n이 짝수라면 n 이하의 짝수인 모든 양의 정수의 제곱의 합을 반환하는 함수를 작성해 주세요. 그 함수가 제대로 작동하는지 테스트하는 프로그램을 작성하세요.

In [ ]:
def sum_odd_or_even(n):
  if n % 2 == 0:
    return sum([i ** 2 for i in range(2, n + 1, 2)])
  else:
    return sum([i for i in range(1, n + 1, 2)])

n = int(input("양의 정수 n을 입력하세요 : "))
print(f"결과는 {sum_odd_or_even(n)}입니다.")

양의 정수 n을 입력하세요 : 10
결과는 220입니다.


Day 30
> 2자리 이상의 정수 number가 사용자의 입력으로 주어집니다.  이 수를 2자리씩 자른 뒤, 자른 수를 모두 더해서 그 합을 출력하는 프로그램을 작성하세요.

>입출력 예제
>
>입력 : 4859
>
>출력 : 107
>
>입력된 수를 2자리씩 나눠 합치면 다음과 같습니다.
>
>48 + 59 = 107

In [ ]:
n = input("2자리 이상의 정수를 입력하세요 : ")
total = 0
for i in range(0, len(n), 2):
  total += int(n[i:i+2])
print(total)

2자리 이상의 정수를 입력하세요 : 4859
107


Day 31
> 자연수 n을 뒤집어 각 자리 숫자를 원소로 가지는 배열 형태로 반환하는 함수를 작성해주세요. 예를 들어 매개변수 n이 12345이면 [5,4,3,2,1]을 반환합니다. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.

In [ ]:
n = int(input("자연수 n을 입력하세요 : "))
print(list(map(int, list(str(n)[::-1]))))

자연수 n을 입력하세요 : 12345
[5, 4, 3, 2, 1]


Day 32
> 매개변수로 주어진 정수를 저장한 리스트 numbers에서 가장 작은 수를 제거한 리스트를 반환하는 함수를 작성해주세요. 단, 반환하려는 리스트가 빈 리스트인 경우엔 리스트에 -1을 채워 반환하세요. 예를 들어 numbers이 [4,3,2,1]인 경우는 [4,3,2]를 리턴 하고, [10]면 [-1]을 반환합니다. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.

In [ ]:
numbers = list(map(int, input("여러 개 수를 공백을 포함하여 입력하세요 : ").split()))

min_n = min(numbers)
numbers.remove(min_n)

if not numbers:
  numbers.append(-1)

print(numbers)

여러 개 수를 공백을 포함하여 입력하세요 : 10
[-1]


Day 33
> 정수 리스트 numbers가 매개변수로 주어집니다. numbers에서 서로 다른 인덱스에 있는 두 개의 수를 뽑아 더해서 만들 수 있는 모든 수를 리스트에 오름차순으로 담아 반환하도록 함수를 완성해주세요. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.
다음과 같은 제한사항이 있습니다.
> 1. numbers의 길이는 2 이상 100 이하입니다.
> 2. numbers의 모든 수는 0 이상 100 이하입니다.

In [ ]:
numbers = list(map(int, input("정수 리스트를 입력하세요 : ").split()))

result = []
for i in range(len(numbers)):
  for j in range(i + 1, len(numbers)):
    result.append(numbers[i] + numbers[j])

print(sorted(list(set(result))))

정수 리스트를 입력하세요 : 1 2 3 4 5 6 7 8 9 10
[3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]


Day 33
> 특정 날짜를 입력하면 그 날짜까지 남은 날이 얼마인지 알려주는 파이썬 코드를 작성해보세요.

In [ ]:
from datetime import datetime, timedelta, timezone

input_date = input("특정 날짜를 입력하세요 (YYYY/MM/DD) : ")
date = datetime.strptime(input_date, "%Y/%m/%d")

# 한국 표준시로 변경
korea_timezone = timezone(timedelta(hours=9))
korea_now = datetime.now(korea_timezone)

date = date.replace(tzinfo=korea_timezone)

# 입력한 날짜와 현재 날짜의 시간 부분을 모두 자정으로 맞추기
date = date.replace(tzinfo=korea_timezone, hour=0, minute=0, second=0, microsecond=0)
korea_now = korea_now.replace(hour=0, minute=0, second=0, microsecond=0)

diff = date - korea_now

print(f"{input_date}까지 남은 날은 {diff.days}일 입니다.")

특정 날짜를 입력하세요 (YYYY/MM/DD) : 2024/12/06
2024-12-06 00:00:00+09:00 2024-10-31 00:00:00+09:00
2024/12/06까지 남은 날은 36일 입니다.


Day 34
> 어떤 문장의 각 알파벳을 일정한 거리만큼 밀어서 다른 알파벳으로 바꾸는 암호화 방식을 "시저 암호(Caesar cipher)"라고 합니다. 예를 들어 "AB"는 1만큼 밀면 "BC"가 되고, 3만큼 밀면 "DE"가 됩니다. "z"는 1만큼 밀면 "a"가 됩니다. 매개변수로 문자열 s와 거리 n을 입력으로 받아 s를 n만큼 민 암호문을 만드는 함수를 완성해주세요. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.

> 다음과 같은 제한 조건을 가집니다.
- 공백은 아무리 밀어도 공백입니다.
- s는 알파벳 소문자, 대문자, 공백으로만 이루어져 있습니다.
- s의 길이는 8000 이하입니다.
- n은 1 이상, 25 이하인 자연수입니다.

In [ ]:
def caesar_cipher(s, n):
  result = ''
  for ch in s:
    if ch.isalpha():
      if ch.isupper(): # 대문자인 경우
        result += chr((ord(ch) - ord('A') + n) % 26 + ord('A'))
      else: # 소문자인 경우
        result += chr((ord(ch) - ord('a') + n) % 26 + ord('a'))
    else: # 공백인 경우
      result += ' '
  return result


s = input("문자열을 입력하세요 : ")
n = int(input("거리 n을 입력하세요 : "))
print(caesar_cipher(s, n))

문자열을 입력하세요 : ABC DEF
거리 n을 입력하세요 : 3
DEF GHI


Day 35
>  자연수 n이 매개변수로 주어집니다. n을 x로 나눈 나머지가 1이 되도록 하는 가장 작은 자연수 x를 반환하는 함수를 완성해주세요. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.

In [ ]:
def find_x(n):
  if n == 1:
      return 2  # 1을 2로 나누면 나머지가 1
  if n == 2:
      return -1  # 조건을 만족하는 x가 존재하지 않음

  # n%x == 1을 푸는 문제는 n-1의 약수를 찾는 문제와 동일
  for x in range(2, int((n - 1) ** 0.5) + 1):
    if (n - 1) % x == 0:
        return x
  # 약수가 없는 경우 n-1이 답
  return n - 1

n = int(input("자연수 n을 입력하세요 : "))
result = find_x(n)
if result != -1:
    print(result)
else:
    print("조건을 만족하는 x가 존재하지 않습니다.")

자연수 n을 입력하세요 : 12
11


Day 36
> 문자열 s는 한 개 이상의 단어로 구성되어 있습니다. 각 단어는 하나 이상의 공백문자로 구분되어 있습니다. 각 단어의 짝수번째 알파벳은 대문자로, 홀수번째 알파벳은 소문자로 바꾼 문자열을 리턴하는 함수를 완성하세요. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.
>
> 제한 사항
- 문자열 전체의 짝/홀수 인덱스가 아니라, 단어(공백을 기준)별로 짝/홀수 인덱스를 판단해야합니다.
- 첫 번째 글자는 0번째 인덱스로 보아 짝수번째 알파벳으로 처리해야 합니다.


In [ ]:
def transform_string(s):
    result = ''
    for word in s.split():
        for i in range(len(word)):
            if i % 2 == 0:
                result += word[i].upper()
            else:
                result += word[i].lower()
        result += ' '  # 각 단어 뒤에 공백 추가

    return result.strip()  # 마지막 공백 제거 후 반환

s = input("문자열을 입력하세요: ")
result = transform_string(s)
print(result)

문자열을 입력하세요: I love you
I LoVe YoU


Day 37
> 매개변수로 정수 n을 입력으로 받아 n의 약수를 모두 더한 값을 반환하는 함수를 완성해주세요.  이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.

In [ ]:
def sum_primes(n):
  total = 0
  for i in range(1, int(n ** 0.5) + 1):
    if n % i == 0:
      total += i
      if i != n // i:
        total += n // i
  return total

n = int(input("자연수 n을 입력하세요 : "))
print(sum_primes(n))

자연수 n을 입력하세요 : 10
18


Day 38
> 정수로 이루어진 배열 numbers가 있습니다. 배열 의 각 원소들에 대해 자신보다 뒤에 있는 숫자 중에서 자신보다 크면서 가장 가까이 있는 수를 뒷 큰수라고 합니다.
>
> 정수 배열 numbers가 매개변수로 주어질 때, 모든 원소에 대한 뒷 큰수들을 차례로 담은 배열을 반환하도록 함수를 완성해주세요. 단, 뒷 큰수가 존재하지 않는 원소는 -1을 담습니다.
이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.
제한사항은 다음과 같습니다.
>
> 4 ≤ numbers의 길이 ≤ 1,000,000
>
> 1 ≤ numbers[i] ≤ 1,000,000
>
> 입출력의 예는 다음과 같습니다.
>
> numbers             result
>
> [2, 3, 3, 5]           [3, 5, 5, -1]
>
> [9, 1, 5, 3, 6, 2]   [-1, 5, 6, 6, -1, -1]
>
> 입출력 예 #1
2의 뒷 큰수는 3입니다. 첫 번째 3의 뒷 큰수는 5입니다. 두 번째 3 또한 마찬가지입니다. 5는 뒷 큰수가 없으므로 -1입니다. 위 수들을 차례대로 배열에 담으면 [3, 5, 5, -1]이 됩니다.
>
> 입출력 예 #2
9는 뒷 큰수가 없으므로 -1입니다. 1의 뒷 큰수는 5이며, 5와 3의 뒷 큰수는 6입니다. 6과 2는 뒷 큰수가 없으므로 -1입니다. 위 수들을 차례대로 배열에 담으면 [-1, 5, 6, 6, -1, -1]이 됩니다

In [ ]:
def find_next_greater_element(numbers):
  result = []
  for i in range(len(numbers)):
    for j in range(i + 1, len(numbers)):
      if numbers[i] < numbers[j]:
        result.append(numbers[j])
        break
    else:
      result.append(-1)
  return result

numbers = list(map(int, input("정수 리스트를 입력하세요 : ").split()))
print(find_next_greater_element(numbers))

정수 리스트를 입력하세요 : 9 1 5 3 6 2
[-1, 5, 6, 6, -1, -1]


Day 39
> 0 또는 양의 정수가 여러 개 주어졌을 때, 정수를 이어 붙여 만들 수 있는 가장 큰 수를 찾아주세요. 예를 들어, 주어진 정수가 [6, 10, 2]라면 [6102, 6210, 1062, 1026, 2610, 2106]를 만들 수 있고, 이중 가장 큰 수는 6210입니다. 0 또는 양의 정수가 담긴 배열 numbers가 매개변수로 주어질 때, 순서를 재배치하여 만들 수 있는 가장 큰 수를 문자열로 바꾸어 반환하도록 함수를 작성해주세요. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.
> 제한사항은 다음과 같습니다.
>
> - numbers의 길이는 1 이상 100,000 이하입니다.
> - numbers의 원소는 0 이상 1,000 이하입니다.

In [ ]:
def largest_number(numbers):
    # 각 숫자를 3번씩 이어 붙이고 sort해서 내림차순 정렬(numbers가 최대 3자리이므로)
    # 예) (입력) ['997', 9'] -> ['997997997', '999'] 비교 '999' > '997997997' -> (출력) ['9', '997']
    # numbers가 10^5 이기 때문에 다른 알고리즘의 경우 오래걸릴 수 있음
    # 단, 1000의 경우는 4자리이지만 영향을 미치지 않음
    sorted_numbers = sorted(numbers, key=lambda x: x * 3, reverse=True)

    largest_num = ''.join(sorted_numbers) # 정렬된 숫자를 합침

    # [0, 0, 0]과 같이 가장 앞자리가 0인 경우 0 출력
    if largest_num[0] == '0':
      return 0
    else:
      return largest_num

numbers = input("정수 리스트를 입력하세요 : ").split()
print(largest_number(numbers))

정수 리스트를 입력하세요 : 6 10 2
6210


Day 40
> 초 단위로 기록된 주식가격이 담긴 배열(리스트) prices가 매개변수로 주어질 때, 가격이 떨어지지 않은 기간은 몇 초인지를 배열(리스트)로 반환하도록 함수를 완성하세요. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.
> 제한사항은 다음과 같습니다.
- prices의 각 가격은 1 이상 10,000 이하인 자연수입니다.
- prices의 길이는 2 이상 100,000 이하입니다.
>
>
>  입출력 예
- prices : [1, 2, 3, 2, 3]
- return : [4, 3, 1, 1, 0]
>
>입출력 예 설명
- 1초 시점의 ₩1은 끝까지 가격이 떨어지지 않았습니다.
- 2초 시점의 ₩2은 끝까지 가격이 떨어지지 않았습니다.
- 3초 시점의 ₩3은 1초뒤에 가격이 떨어집니다. 따라서 1초간 가격이 떨어지지 않은 것으로 봅니다.
- 4초 시점의 ₩2은 1초간 가격이 떨어지지 않았습니다.
- 5초 시점의 ₩3은 0초간 가격이 떨어지지 않았습니다

In [ ]:
def calc_price_time(prices):
  result = []

  for i, price in enumerate(prices):
    time = 0
    for j in range(i + 1, len(prices)):
      time += 1
      if price > prices[j]:
        break
    result.append(time)

  return result

prices = list(map(int, input("주식가격 리스트를 입력하세요 : ").split()))
print(calc_price_time(prices))

주식가격 리스트를 입력하세요 : 1 2 3 2 3
[4, 3, 1, 1, 0]


Day 41
> 리스트 안에 리스트가 중첩되어있는 아래와 같은 중첩 리스트(nested list)를
- [[1, 2, [3]], [4, [5, 6]], 7, [8, 9]]
- 다음과 같은 1차원 리스트로 평탄화하는 방법을 찾아보세요.
- [1, 2, 3, 4, 5, 6, 7, 8, 9]

In [ ]:
def flatten_nested_list(nested_list):
  if isinstance(nested_list, list):
    result = []
    for item in nested_list:
      result.extend(flatten_nested_list(item))
    return result
  return [nested_list]

nested_list = [[1, 2, [3]], [4, [5, 6]], 7, [8, 9]]
print(flatten_nested_list(nested_list))

[1, 2, 3, 4, 5, 6, 7, 8, 9]


Day 42
> 양의 정수 x가 하샤드 수이려면 x의 자릿수의 합으로 x가 나누어져야 합니다. 예를 들어 18의 자릿수 합은 1+8=9이고, 18은 9로 나누어 떨어지므로 18은 하샤드 수입니다. 자연수 x를 매개변수로 입력받아 x가 하샤드 수인지 아닌지 검사하는 함수를 완성해주세요. 제한 조건은 x는 1 이상, 10000 이하인 정수입니다.

In [ ]:
x = input("자연수를 입력하세요 : ")
x_sum = sum(map(int, x))

if int(x) % x_sum == 0:
  print("하샤드 수입니다.")
else:
  print("하샤드 수가 아닙니다.")

자연수를 입력하세요 : 18
하샤드 수입니다.


Day 43
> 2개의 문자열을 받아서 이들 문자열에 모두 포함된 글자를 반환하는 프로그램을 작성해보자.

In [ ]:
s1 = input('첫 번쨰 문자열을 입력하세요: ')
s2 = input('두 번째 문자열을 입력하세요: ')
print(f'2개의 문자열에 모두 포함된 글자 : {set(s1) & set(s2)}')

첫 번쨰 문자열을 입력하세요: taxi
두 번째 문자열을 입력하세요: tai
2개의 문자열에 모두 포함된 글자 : {'t', 'a', 'i'}


Day 44
> 딕셔너리에 쇼핑몰에서 구입한 상품의 가격이 저장되어 있다. 딕셔너리에 있는 모든 상품 가격의 합계를 계산하는 프로그램을 작성해보자.

In [ ]:
dict_ = {'물' : 700, '주스':1300, '과자': 2000, '사탕' : 200, '초콜릿': 1300}

total = 0
for product in dict_:
  total += dict_[product]

print(f'모든 상품의 가격의 합계 : {total}원')

모든 상품의 가격의 합계 : 5500원


Day 45
> 키와 값을 저장하고 있는 2개의 리스트를 합쳐서 하나의 딕셔너리로 만드는 프로그램을 작성해보세요. enumerate() 함수를 사용해보세요

In [ ]:
keys = ['물', '주스', '과자', '사탕', '초콜릿']
values = [700, 1300, 2000, 200, 1300]

dict_ = {key:values[i] for i, key in enumerate(keys)}
print(dict_)

{'물': 700, '주스': 1300, '과자': 2000, '사탕': 200, '초콜릿': 1300}


Day 46
> 달의 번호를 받아서 달의 이름을 출력하는 프로그램을 작성해보자. 예를 들어서 1월이 입력되면 "January"가 출력된다. 이것은 물론 if-else 구문을 사용해서도 구현이 가능하다. 파이썬이 제공하는 데이터 타입을 사용해서 이 기능을 구현해보자.

In [ ]:
months = { 1: "January", 2: "February", 3: "March", 4: "April", 5: "May", 6: "June", 7: "July", 8: "August", 9: "September", 10: "October", 11: "November", 12: "December"}

month = int(input("달의 번호를 입력하세요 : "))
if month in months:
  print(f"{month}월은  {months[month]}입니다.")
else:
  print('잘못된 입력입니다.')

달의 번호를 입력하세요 : 1
1월은  January입니다.


Day 47
> "MM/DD/YYYY" 형식으로 표기된 날짜를 매개변수로 받아서 "YYYYMMDD" 형식으로 반환하는 함수를 작성해보자. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요

In [ ]:
dates = input("날짜를 입력하세요 (MM/DD/YYYY) : ").split('/')
print(f"{dates[2]}{dates[0]}{dates[1]}")

날짜를 입력하세요 (MM/DD/YYYY) : 12/06/2024
20241206


Day 48
> 어떤 정수들이 있습니다. 이 정수들의 절댓값을 차례대로 담은 정수 리스트 absolutes와 이 정수들의 부호를 차례대로 담은 Boolean 리스트 signs가 매개변수로 주어집니다. 실제 정수들의 합을 구하여 반환하도록 함수를 작성해보자. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.
다음과 같은 제한사항이 있다.
- absolutes의 길이는 1 이상 1,000 이하입니다.
  - absolutes의 모든 수는 각각 1 이상 1,000 이하입니다.
- signs의 길이는 absolutes의 길이와 같습니다.
  -  signs[i] 가 참이면 absolutes[i] 의 실제 정수가 양수임을, 그렇지 않으면 음수임을 의미합니다.

In [ ]:
def real_sum(absolutes, signs):
  total = 0
  for n, sign in zip(absolutes, signs):
    if sign == False:
      total += -1 * n
    else:
      total += n
  return total

absolutes = [5, 3, 4, 7, 2, 8]
signs = [False, True, True, False, True, False]
print(real_sum(absolutes, signs))

-11


Day 49
> The last rose of summer 라는 시를 저장하고 있는 텍스트 파일이 있다고 가정하자.
이 텍스트 파일을 읽어서 가장 긴 단어를 찾는 프로그램을 작성해보자.
>
>
>`Tis the last rose of Summer,
   Left blooming alone;
All her lovely companions
   Are faded and gone;
No flower of her kindred,
   No rose-bud is nigh,
To reflect back her blushes
>  Or give sigh for sigh!
>
> I'll not leave thee, thou lone one,
   To pine on the stem;
Since the lovely are sleeping,
   Go sleep thou with them.
Thus kindly I scatter
   Thy leaves o'er the bed
Where thy mates of the garden
>   Lie scentless and dead.
>
> So soon may I follow,
   When friendships decay,
And from Love's shining circle
   The gems drop away!
When true hearts lie withered,
   And fond ones are flown,
Oh! who would inhabit
   This bleak world alone?

In [ ]:
with open('The last rose of summer.txt', 'r') as f:
  longest_word = ''
  for line in f:
    for word in line.split():
      word = word.strip('\'\",.;!?')
      if len(word) > len(longest_word):
        longest_word = word

print(f'가장 긴 단어는 {longest_word} 입니다.')

가장 긴 단어는 friendships 입니다.


Day 50
> 현재의 작업 디렉토리에서 "numbers.txt" 파일을 열어서 1부터 100사이의 10개의 정수 난수를 문자열 형태로 추가하고 파일을 닫는 프로그램을 작성하세요.

In [ ]:
import random

numbers = random.sample(range(1, 101), k = 10)

with open('numbers.txt', 'a') as f:
    for number in numbers:
        f.write(str(number) + '\n')

Day 51
> 학생들의 성적이 부동소수점 수로 파일 scores.txt에 저장되어 있다고 하자(메모장에서 ANSI 엔코딩으로 저장한다).
이 성적을 읽어서 파일의 끝에 평균값을 추가하는 프로그램을 작성하세요.
> - 99.1
> - 88.2
> - 67.7
> - 96.9

In [ ]:
n_list = []
with open('scores.txt', 'r+') as f:
  while True:
    n = f.readline()
    if not n:
        break
    n_list.append(float(n))

  avg_score = sum(n_list) / len(n_list)
  print(f'평균 값은 {avg_score}')
  f.write('\n' + str(avg_score))

평균 값은 87.975


Day 52
> 메모장 소프트웨어를 이용하여 아래의 시를 텍스트 파일로 만드세요.
그 텍스트 파일을 열어 각 줄의 앞에 번호를 매겨서 다시 파일에 쓰는 프로그램을 작성해보세요.
>
```
Wine comes in at the mouth
And love comes in at the eye;
That's all we shall know for truth
Before we grow old and die.
I lift the glass to my mouth, I look at you, and I sigh.
```


In [ ]:
lines = []
with open('text.txt', 'r') as f:
  while True:
    line = f.readline()
    if not line:
      break
    lines.append(line)

with open('text.txt', 'w') as f:
  for i, line in enumerate(lines):
    f.write(f'{i + 1}. {line}')

Day 52 (2)
> 다음과 같은 가능을 수행하는 파이썬 프로그램을 작성하세요.
> 1. random 모듈의 randint() 메소드를 사용하여 1에서 1,000 까지의 임의의 정수를 10개 생성하여 random_numbers.txt 라는 이름의 파일을 저장하세요.
> 2.random_numbers.txt 라는 파일로부터 숫자를 읽어 들인 후 이들 중 짝수 값을 가지는 정수를 저장하는 random_even.txt 라는 파일을 생성하세요.

In [ ]:
import random

n_list = [random.randint(1, 1001) for n in range(10)]

with open('random_numbers.txt', 'w') as f:
  for n in n_list:
    f.write(str(n) + '\n')

with open('random_numbers.txt', 'r') as f, open('random_even.txt', 'w') as f2:
  while True:
    n = f.readline()
    if not n:
      break
    if int(n) % 2 == 0:
      f2.write(n)

Day 53
> random 모듈에 있는 함수를 사용하여 알파벳 a 부터 z 사이에서 랜덤하게 10개의 문자를 출력하는 프로그램을 작성해보세요.

In [ ]:
import random
import string

alphabets = random.sample(string.ascii_lowercase, 10)
print(alphabets)

['a', 'f', 'm', 'g', 'j', 'n', 'l', 'e', 'u', 'p']


Day 54
> random 모듈에 있는 함수를 사용하여 가위바위보 게임을 작성해보세요. 사용자와 여러분이 만든 프로그램이 가위바위보 게임을 하도록 구현해보세요.

In [ ]:
import random

user_choice  = input('가위, 바위, 보 중 하나를 선택하세요 : ')

if user_choice  not in ['가위', '바위', '보']:
  print('잘못된 입력입니다.')
else:
  computer_choice = random.choice(['가위', '바위', '보'])
  print(f'컴퓨터의 선택 : {computer_choice}')
  if user_choice == computer_choice:
    print('비겼습니다.')
  elif (user_choice == '가위' and computer_choice == '보') or (user_choice == '바위' and computer_choice == '가위') or (user_choice == '보' and computer_choice == '바위'):
    print('사용자가 이겼습니다.')
  else:
    print('사용자가 졌습니다.')

가위, 바위, 보 중 하나를 선택하세요 : 가위
컴퓨터의 선택 : 보
사용자가 이겼습니다.


Day 55
> 다음 파이썬 연산을` __add__(), __sub__(), __mul__(), __truediv__()` 함수를 사용하여 구현하세요.
>
> - (1) 123 + 334
> - (2) 123 - 334
> - (3) 123 * 334
> - (4) 123 / 3

In [ ]:
print((123).__add__(334))
print((123).__sub__(334))
print((123).__mul__(334))
print((123).__truediv__(3))

457
-211
41082
41.0


Day 56

> 고양이를 클래스로 정의하고 몇 개의 인스턴스를 생성해보자. 접근자와 설정자를 사용해보자.
> 고양이 클래스에는 다음과 같은 인스턴스 변수들이 있다.
>    - name : 고양이의 이름
     - age : 고양이의 나이
>
> 고양이 클래스에는 다음과 같은 메소드들이 있다.
>    - `__init__(self, name, age)` : 생성자 함수
>    - `__str__()` : 고양이의 정보를 문자열로 반환하는 함수
>
>    - `setName(), getName(), setAge(), getAge()` : 각 속성에 대한 접근자와 설정자 함수들

In [ ]:
class Cat:
  def __init__(self, name, age):
    self.__name = name
    self.__age = age

  def __str__(self):
    return f'이름 : {self.__name}, 나이 : {self.__age}'

  def setName(self, name):
    self.__name = name

  def getName(self):
    return self.__name

  def setAge(self, age):
    self.__age = age

  def getAge(self):
    return self.__age

nabi = Cat('나비', 100)
nabi.setName('나비야')
nabi.setAge(10)
print(nabi.getName())
print(nabi.getAge())
print(nabi)

나비야
10
이름 : 나비야, 나이 : 10


Day 57
> 사각형을 나타내는 Rectangle 클래스를 작성해보자.
> Rectangle 클래스는 다음과 같은 인스턴스 변수와 메소드를 가진다.
두 개의 사각형 객체를 생성해보고, 서로 겹치는지 확인해보자.
>
> 인스턴스 변수
>
>    - x, y : 사각형의 좌측 상단 좌표
>    - width, height : 사각형의 너비와 높이
>
> 메소드
>    - `__init__(self, x, y, w, h)` : 매개변수 x, y, w, h를 가지는 생성자 함수
>    - `__str__()` : 사각형의 좌표와 크기를 문자열로 반환하는 함수
>    - `getArea() `: 사각형의 면적을 계산하여 반환한다.
>    - `overlap(r)` : 현재 사각형과 전달된 사각형이 겹치면 True, 그렇지 않으면 False를 반환한다

In [ ]:
class Rectangle:
  def __init__(self, x, y, width, height):
    self.x = x
    self.y = y
    self.width = width
    self.height = height

  def __str__(self):
    return f'Rectange(x = {self.x}, y = {self.y}, width = {self.width}, height = {self.height})'

  def getArea(self):
    return self.width * self.height

  def overlap(self, r):
    if  (self.x + self.width <= r.x or  # self가 r의 왼쪽에 있음
          self.x >= r.x + r.width or    # self가 r의 오른쪽에 있음
          self.y + self.height <= r.y or  # self가 r의 아래쪽에 있음
          self.y >= r.y + r.height):  # self가 r의 위쪽에 있음
      return False
    else:
      return True

rect1 = Rectangle(0, 0, 4, 4)
rect2 = Rectangle(2, 2, 4, 4)
rect3 = Rectangle(5, 5, 2, 2)

print(rect1)
print(rect2)
print(rect3)
print()

print(rect1.getArea())
print(rect2.getArea())
print(rect3.getArea())
print()

print(rect1.overlap(rect2))
print(rect1.overlap(rect3))
print(rect2.overlap(rect3))

Day 58
>사람들의 연락처를 저장하는 PhoneBook 클래스를 작성해보자.
>PhoneBook 클래스는 딕셔너리를 이용하여서 연락처를 저장한다.
>PhoneBook 객체를 2개 생성해보고 객체를 출력해보자.
>
> 인스턴스 변수
>   -  contacts : 연락처를 저장하는 딕셔너리, 연락처는 name, mobile, email 로 구성된다.
>
> 메소드
>    - `__init__(self)` : 생성자 함수, 생성자에서 딕셔너리를 생성한다.
>    - `__str__()` : 연락처 정보를 문자열로 변환하는 함수
>    - `add(self, name, mobile=None, email=None)` : 한 사람의 연락처를 추가하는 함수

In [ ]:
class PhoneBook:
  def __init__(self):
    self.contacts = {}

  def __str__(self):
    return str(self.contacts)

  def add(self, name, mobile=None, email=None):
    self.contacts[name] = {'mobile': mobile, 'email': email}

pb1 = PhoneBook()
pb2 = PhoneBook()

pb1.add('홍길동', '010-1111-1111', 'gildong@naver.com')
pb1.add('김길동', '010-2222-2222', 'kim@naver.com')
pb2.add('박길동', '010-3333-3333', 'park@naver.com')
pb2.add('이길동', '010-4444-4444', 'lee@naver.com')

print(pb1)
print(pb2)

{'홍길동': {'mobile': '010-1111-1111', 'email': 'gildong@naver.com'}, '김길동': {'mobile': '010-2222-2222', 'email': 'kim@naver.com'}}
{'박길동': {'mobile': '010-3333-3333', 'email': 'park@naver.com'}, '이길동': {'mobile': '010-4444-4444', 'email': 'lee@naver.com'}}


Day 59
> 다음과 같이 실행할 때 입력 값을 모두 더하여 출력하는 스클립트(c:\pythonscripts\myargv.py)를 작성해보자.
>
> - c:\> cd pythonscripts
> - c:\pythonscripts> python myargv.py 1 2 3 4 5 6 7 8 9 10

In [ ]:
import argparse

parser = argparse.ArgumentParser()
parser.add_argument('numbers', type=int, nargs='+')
args = parser.parse_args()

print(sum(args.numbers))

Day 60
> time 모듈을 사용하여 현재 날짜와 시간을 다음과 같은 형식으로 출력해보자.
> - 2024/11/28 09:00:00

In [ ]:
import time

current_time = time.strftime("%Y/%m/%d %H:%M:%S")

print(current_time)

2024/12/07 15:37:11


Day 61
> 어떤 튜플이 주어졌을 때, 처음에는 튜플 전체를 출력하여라.
그 다음에는 마지막 한 요소를 제외하고 출력하며,
그 다음 단계에는 마지막 두 요소를 제외하고 출력하는 코드를 슬라이싱을 이용하여 작성해보세요.
>
아래의 예제를 참고하세요.

```(4, 5, 2, 3, 8, 1, 9, 0)
(4, 5, 2, 3, 8, 1, 9)
(4, 5, 2, 3, 8, 1)
(4, 5, 2, 3, 8)
(4, 5, 2, 3)
(4, 5, 2)
(4, 5)
(4,)
```

In [ ]:
tup = (4, 5, 2, 3, 8, 1, 9)

n = len(tup)
for i in range(n):
  print(tup[:n - i])

(4, 5, 2, 3, 8, 1, 9)
(4, 5, 2, 3, 8, 1)
(4, 5, 2, 3, 8)
(4, 5, 2, 3)
(4, 5, 2)
(4, 5)
(4,)


Day 62
> 다음과 같은 리스트가 있다.
> - numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
>
> numbers 리스트의 요소 중에서 홀수에만 2를 곱하여 새로운 리스트 result를 만들어보자.
>
> 이때 리스트 축약(list comprehension)을 사용해보자.

In [ ]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

print([n * 2 for n in numbers if n % 2 == 1])

[2, 6, 10, 14, 18]


Day 63
> 다음과 같은 2개의 리스트가 있다.
2개의 리스트에 공통적인 요소를 추출하는 리스트 축약(list comprehension) 코드를 작성해보자.
```
a = [1, 2, 3, 4, 5]
b = [1, 3, 3, 4, 5, 6, 7, 8]
```

In [ ]:
a = [1, 2, 3, 4, 5]
b = [1, 3, 3, 4, 5, 6, 7, 8]

print([n for n in a if n in b])

[1, 3, 4, 5]


Day 64
> 2개의 리스트를 매개변수로 받아서 공통 구성원이 하나 이상 있으면 True를 반환하는 함수를 작성하고 테스트해보자.

In [ ]:
def common_elements(a, b):
  return True if set(a) & set(b) else False

a = [1, 2, 3]
b = [7, 1, 5]
print(common_elements(a, b))

True


Day 65
> 람다 함수를 사용하여 정수 리스트를 필터링하는 프로그램을 작성해보자.
- 원래 리스트 : [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
- 짝수 리스트 : [2, 4, 6, 8, 10]
- 홀수 리스트 : [1, 3, 5, 7, 9]

In [ ]:
n_list = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
even_list = list(filter(lambda x : x%2 == 0 , n_list))
odd_list = list(filter(lambda x : x%2 == 1, n_list))

print(f'원래 리스트 : {n_list}')
print(f'짝수 리스트 : {even_list}')
print(f'홀수 리스트 : {odd_list}')

원래 리스트 : [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
짝수 리스트 : [2, 4, 6, 8, 10]
홀수 리스트 : [1, 3, 5, 7, 9]


Day 66
- 없음

Day 67
> map()와 filter()를 사용하여서 [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] 중에서 짝수의 제곱으로 새로운 리스트를 만드는 프로그램을 작성해보세요.

In [ ]:
print(list(map(lambda x : x ** 2, filter(lambda x : x % 2 == 0, [1,2,3,4,5,6,7,8,9,10]))))

[4, 16, 36, 64, 100]


Day 68
> random 모듈에 있는 함수를 사용하여서 알파벳 a 부터 z 사이에서 랜덤하게 10개의 문자를 출력하는 프로그램을 작성해보세요.

In [ ]:
from string import ascii_lowercase
import random

print(random.sample(ascii_lowercase, 10))

['r', 'w', 'g', 'f', 'u', 'c', 'd', 'v', 'm', 'p']


Day 69
> Python one-line code 샘플을 찾아보고, 여러분께서 이해한 바를 기술해보세요.

관련 링크 : https://wiki.python.org/moin/Powerful%20Python%20One-Liners

one-line code는 코드를 간결하고, 가독성을 높게 해주는 방법이다. list comprehension의 경우에는 메모리 및 시간 효율성까지 높일 수 있다.

다만, 너무 복잡한 경우 가독성이 저하될 수 있다.
```
result = [x*y for x in range(1, 5) for y in range(1, 5) if x != y]
```
-> 1부터 4까지의 숫자 쌍 (x, y) 중에서 x와 y가 서로 다른 경우에 대해 x * y 값을 구하여 결과 리스트를 만드는 코드

때문에 one-line code를 사용할 때에는 간단한 논리와 작업에는 효율적이지만, 복잡한 작업에서는 가독성과 유지보수성을 해치지 않는 범위 내에서 활용하는 것이 중요해보인다.

Day 70
> String, List, Set, Tuple, Dictionary 데이터 타입의 주요 메소드를 선정하고, 사용 예제를 만들어보세요.

In [5]:
# String
text = "  Hello, World!  "

# 소문자로 변환
print(text.lower())

# 대문자로 변환
print(text.upper())

# 공백 제거
print(text.strip())

# 문자열 분리
print(text.split(', '))

# 문자열 교체
print(text.replace("World", "Python"))

  hello, world!  
  HELLO, WORLD!  
Hello, World!
['  Hello', 'World!  ']
  Hello, Python!  


In [8]:
# List
numbers = [3, 1, 4, 1]

# 요소 추가
numbers.append(5)
print(numbers)

# 여러 요소 추가
numbers.extend([6, 2, 7])
print(numbers)


# 특정 위치에 요소 삽입
numbers.insert(3, 10)
print(numbers)

# 특정 요소 제거
numbers.remove(10)
print(numbers)

# 마지막 요소 제거 및 반환
last = numbers.pop()
print(last)
print(numbers)

# 정렬
numbers.sort()
print(numbers)

[3, 1, 4, 1, 5]
[3, 1, 4, 1, 5, 6, 2, 7]
[3, 1, 4, 10, 1, 5, 6, 2, 7]
[3, 1, 4, 1, 5, 6, 2, 7]
7
[3, 1, 4, 1, 5, 6, 2]
[1, 1, 2, 3, 4, 5, 6]


In [9]:
# Set
my_set = {1, 2, 3}

# 요소 추가
my_set.add(4)
print(my_set)

# 요소 제거
my_set.remove(2)
print(my_set)

# 합집합
other_set = {3, 4, 5, 6}
union_set = my_set.union(other_set)
print(union_set)

# 교집합
intersection_set = my_set.intersection(other_set)
print(intersection_set)

# 차집합
difference_set = my_set.difference(other_set)
print(difference_set)

{1, 2, 3, 4}
{1, 3, 4}
{1, 3, 4, 5, 6}
{3, 4}
{1}


In [10]:
# Tuple
my_tuple = (1, 2, 3, 2, 4, 2)

# 특정 값의 개수 세기
count_of_2 = my_tuple.count(2)
print(count_of_2)

# 특정 값의 첫 번째 인덱스 찾기
index_of_3 = my_tuple.index(3)
print(index_of_3)

3
2


In [12]:
# Dict
my_dict = {'a' : 1, 'b' :  2, 'c' : 3}

# key에 해당하는 value 찾기
print(my_dict.get('a'))
print(my_dict.get('d')) # None

# 모든 키 가져오기
print(my_dict.keys())

# 모든 값 가져오기
print(my_dict.values())

# 키와 값 쌍 가져오기
print(my_dict.items())

# 다른 딕셔너리와 병합
my_dict.update({'d': 4, 'e' : 5})
print(my_dict)

1
None
dict_keys(['a', 'b', 'c'])
dict_values([1, 2, 3])
dict_items([('a', 1), ('b', 2), ('c', 3)])
{'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5}


Day 71
> 화씨 온도 데이터를 저장하고 있는 리스트가 있다. 람다 함수와 map( ) 함수를 사용하여 섭씨 온도 데이터로 변환한 리스트의 값을 출력하세요.

In [13]:
fahrenheit = [41, 32, 212]

celsius = list(map(lambda x : (x - 32) * 5/9, fahrenheit))

print(celsius)

[5.0, 0.0, 100.0]


Day 72
> 강의를 하면서 몇 개의 두문자어(頭文字語, 예 FYI(For Your Infromation))를 소개해드렸습니다.
어느 산업 분야에서 일을 하시던지 그 곳의 jargon, terminology를 빠르게 익힐 수 있도록 노력하세요.
사용자가 문장을 입력하면 해당되는 두문자어를 출력하는 프로그램을 작성해보세요.

In [15]:
sentence = input('문장을 입력하세요 : ')
initialism = map(lambda x : x[0], sentence.split())
print("".join(initialism))

문장을 입력하세요 : For Your Information
FYI


Day 73
> 특정 웹 페이지의 HTML 소스 코드를 확인하고, 파일로 저장해보자.
requests 모듈의 get( ) 함수는 웹 서버에 GET 요청을 보내고, 웹 서버가 보내주는 응답 객체를 받을 수 있다.
HTML 소스 코드의 경우 응답 객체의 text 속성에 저장되어 있다.

In [22]:
import requests
url = "https://www.naver.com"
response = requests.get(url)

print("HTML 소스 코드")
print(response.text)

HTML 소스 코드
   <!doctype html> <html lang="ko" class="fzoom"> <head> <meta charset="utf-8"> <meta name="Referrer" content="origin"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <meta name="viewport" content="width=1190"> <title>NAVER</title> <meta name="apple-mobile-web-app-title" content="NAVER"/> <meta name="robots" content="index,nofollow"/> <meta name="description" content="네이버 메인에서 다양한 정보와 유용한 컨텐츠를 만나 보세요"/> <meta property="og:title" content="네이버"> <meta property="og:url" content="https://www.naver.com/"> <meta property="og:image" content="https://s.pstatic.net/static/www/mobile/edit/2016/0705/mobile_212852414260.png"> <meta property="og:description" content="네이버 메인에서 다양한 정보와 유용한 컨텐츠를 만나 보세요"/> <meta name="twitter:card" content="summary"> <meta name="twitter:title" content=""> <meta name="twitter:url" content="https://www.naver.com/"> <meta name="twitter:image" content="https://s.pstatic.net/static/www/mobile/edit/2016/0705/mobile_212852414260.png"> <meta name="twitter:des

# Day 74
> 다음과 같은 리스트가 주어졌다.
>
>list_ = [1, 2, 4, 5]
>
>list_ 리스트에 하나의 원소를 추가하여 다음과 같은 리스트를 만드세요.
>
>list_ = [1, 2, 3, 4, 5]

In [26]:
list_ = [1, 2, 4, 5]
list_.insert(2, 3)
print(list_)

[1, 2, 3, 4, 5]


# Day 75
- 없음